2020/5/boarding_passes/src/main.rs

```diff
@@ -0,0 +1,168 @@
+use std::collections::HashSet;
+use std::env;
+use std::fs::File;
+use std::io::{self, BufRead};
+
+fn main() -> Result<(), Error> {
+    let args: Vec<String> = env::args().collect();
+    let file = File::open(&args[1])?;
+    let reader = io::BufReader::new(file);
+
+    // if let Ok(max) = max_id(reader) {
+    //     println!("The largest id is {}", max);
+    // }
+
+    let mut all_passes = HashSet::new();
+    for line in reader.lines() {
+        if let Ok(l) = line {
+            let bp = BoardingPass::new(&l)?;
+            all_passes.insert(bp.seat_id());
+        }
+    }
+
+    for row in 1..127 {
+        for col in 0..7 {
+            let id = 8 * row + col;
+            if !all_passes.contains(&id)
+                && all_passes.contains(&(id - 1))
+                && all_passes.contains(&(id + 1))
+            {
+                println!("Your seat id is {}", id);
+            }
+        }
+    }
+
+    Ok(())
+}
+
+fn max_id(reader: io::BufReader<std::fs::File>) -> Result<usize, Error> {
+    let mut max = 0;
+    for line in reader.lines() {
+        if let Ok(l) = line {
+            let bp = BoardingPass::new(&l)?;
+            let id = bp.seat_id();
+            if id > max {
+                max = id
+            }
+        }
+    }
+
+    Ok(max)
+}
+
+pub struct BoardingPass {
+    definition: String,
+}
+
+#[derive(Debug)]
+pub enum Error {
+    InvalidBoardingPass,
+    IOError(std::io::Error),
+}
+
+impl std::convert::From<std::io::Error> for Error {
+    fn from(e: std::io::Error) -> Error {
+        Error::IOError(e)
+    }
+}
+
+impl BoardingPass {
+    pub fn new(definition: &str) -> Result<BoardingPass, Error> {
+        if definition.len() == 10
+            && chars_in(&definition[0..7], &['F', 'B'])
+            && chars_in(&definition[8..10], &['L', 'R'])
+        {
+            Ok(BoardingPass {
+                definition: definition.into(),
+            })
+        } else {
+            Err(Error::InvalidBoardingPass)
+        }
+    }
+
+    pub fn row(&self) -> usize {
+        bsp(&self.definition[0..7], 'B', 'F')
+    }
+
+    pub fn col(&self) -> usize {
+        bsp(&self.definition[7..10], 'R', 'L')
+    }
+
+    pub fn seat_id(&self) -> usize {
+        8 * self.row() + self.col()
+    }
+}
+
+fn chars_in(val: &str, allowed: &[char]) -> bool {
+    for c in val.chars() {
+        if !allowed.contains(&c) {
+            return false;
+        }
+    }
+
+    true
+}
+
+fn bsp(val: &str, up: char, down: char) -> usize {
```

Comment:
Wanna know something spicy I realized after solving the puzzle this way? If you treat `F` and `L` as `0` and `B` and `R` as `1` the boarding pass code is a binary encoding of the seat ID 🤯.

Reply:
Wow. Makes sense, but I totally didn't come up with that either.

Reply:
I only got it after seeing other solutions (we've got a Slack channel where people are sharing solutions) :). Cool though!

Reply:
https://github.com/anurse/advent-of-code/commit/6488f31d7fb0832fe7e228381a24ed20e37a5fc4

So much simpler!